SteamKit2/SteamKit2/Networking/Steam3/WebSocketConnection.cs

```diff
@@ -0,0 +1,226 @@
+﻿using System;
+using System.IO;
+using System.Net;
+using System.Net.WebSockets;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace SteamKit2
+{
+    class WebSocketConnection : IConnection
+    {
+        WebSocketContext currentContext;
+
+        public event EventHandler<NetMsgEventArgs> NetMsgReceived;
+
+        public event EventHandler Connected;
+
+        public event EventHandler<DisconnectedEventArgs> Disconnected;
+
+        public EndPoint CurrentEndPoint => currentContext?.EndPoint;
+
+        public ProtocolTypes ProtocolTypes => ProtocolTypes.WebSocket;
+
+        public void Connect(EndPoint endPoint, int timeout = 5000)
+        {
+            if (!(endPoint is DnsEndPoint dnsEp))
+            {
+                DebugLog.WriteLine(nameof(WebSocketConnection), "Given endpoint was not a DnsEndPoint.");
+                Disconnected?.Invoke(this, new DisconnectedEventArgs(false));
+                return;
+            }
+
+            var newContext = new WebSocketContext(this, dnsEp);
+            var oldContext = Interlocked.Exchange(ref currentContext, newContext);
+            if (oldContext != null)
+            {
+                DebugLog.WriteLine(nameof(WebSocketConnection), "Attempted to connect while already connected. Closing old connection...");
+                oldContext.Dispose();
+                Disconnected?.Invoke(this, new DisconnectedEventArgs(false));
+            }
+
+            newContext.Start(TimeSpan.FromMilliseconds(timeout));
+        }
+
+        public void Disconnect()
+            => DisconnectCore(userInitiated: true, specificContext: null);
+
+        public IPAddress GetLocalIP() => IPAddress.None;
+
+        public void Send(byte[] data)
+        {
+            try
+            {
+                currentContext?.SendAsync(data).GetAwaiter().GetResult();
```

Comment:
This doesn't do anything with deadlocks, it just throws the actual exception instead of AggregateException - i.e. it saves me having to catch it, then rethrow the inner exception with `ExceptionDispatchInfo`.